### DISCUSSION TITLE
Future of `LegacyStoreManagers`

### DISCUSSION THREAD

**Discussion by mhajas** [keycloak/keycloak#25898] [Ideas]
Keycloak legacy storage contains several layers (For example for users):
1. Infinispan caching
2. User storage manager
3. JPA layer

Before the `MapStorage` work each layer was available from `KeycloakSession`:
1. https://github.com/keycloak/keycloak/blob/b1d22d6463d6652b26f8b7180b7e9fec265b77de/server-spi/src/main/java/org/keycloak/models/KeycloakSession.java#L136
2. https://github.com/keycloak/keycloak/blob/b1d22d6463d6652b26f8b7180b7e9fec265b77de/server-spi/src/main/java/org/keycloak/models/KeycloakSession.java#L146
3. https://github.com/keycloak/keycloak/blob/b1d22d6463d6652b26f8b7180b7e9fec265b77de/server-spi/src/main/java/org/keycloak/models/KeycloakSession.java#L160

The idea of `MapStorage` was to hide the layering from the services layer (layer using storage) and therefore `KeycloakSession`  was changed to include only one method to access users and the new store was built to count with that. With the removal of `MapStorage` these methods (accessing different storage layers e.g. skipping caching) are still relevant and necessary.

All places relying on the knowledge of the layering were considered deprecated and were supposed to be replaced in future (Old Storage SPI often relies on this functionality). We introduced the following interface containing these methods: https://github.com/keycloak/keycloak/blob/6c4436a76f804ab760ef2e3bc06c6c18c54924da/model/legacy/src/main/java/org/keycloak/storage/LegacyStoreManagers.java#L27 and we use the following method for accessing local storage:
https://github.com/keycloak/keycloak/blob/6c4436a76f804ab760ef2e3bc06c6c18c54924da/model/legacy-private/src/main/java/org/keycloak/storage/UserStoragePrivateUtil.java#L28-L30. This interface was introduced because we planned to remove these methods in the future as `MapStorage` didn't need them.

Currently, this works with the Legacy store but the `UserStoragePrivateUtil` method throws an exception when some custom implementation of `DatastoreProvider` is used and the method is called (it is called for example when LDAP is used). There are more details on this in the following discussion: https://github.com/keycloak/keycloak/discussions/24052#discussioncomment-7966101 and an attempt to fix this: https://github.com/keycloak/keycloak/pull/25572

This attempt to fix would solve the issue and the method would not throw the exception anymore.

However, does it make sense to allow implementing `DatastoreProvider` without implementing methods from `LegacyStoreManagers`? With the `MapStorage` removal these methods are still relevant and not implementing them can result in some unexpected behavior like infinite loops. 

To summarize, there are two solutions:
1. Allow creating `DatastoreProvider`s without `LegacyStoreManagers` and make sure the methods from the interface are called only on providers that implement this interface. 
2. Make `DatastoreProvider` extend `LegacyStoreManager` so each implementation of the `DatastoreProvider` needs to provide a way to access each layer of the storage separately.

**Comment by ahus1** [comment]
My 2 cents on `KeycloakSession`: I know that the comment above doesn't advocate for it, but still: The old methods should not return to `KeycloakSession`, as a regular developers wouldn't know when to call which of the methods. Only those implementing datastore providers should need to know about them. 

I don't have an opinion on 1 vs. 2 yet.

**Comment by dominikschlosser** [comment]
I think solution 2 is the better one, as this allows the custom implementation to decide for itself how to implement the caching layer. Right now we extend `LegacyDatastore` for our Cassandra-extension and simply delegate the manager-methods to the non-manager-methods (`userStorageManager()` just returns `users()`, same for `userLocalStorage()`).
